### ISSUE TITLE
Does Pubsub.topic() make an HTTP request?

### ISSUE THREAD

**Issue by jgeewax** [GoogleCloudPlatform/gcloud-ruby#146]
In reference to http://googlecloudplatform.github.io/gcloud-ruby/docs/master/Gcloud/Pubsub/Project.html#topic-method

Topic itself doesn't have any real data with it, so in the other libraries we've had to think about this pretty carefully... I believe Node does an idea of "auto-create"... 

In short, is there a way to "get" a topic without making an HTTP request? And is there a way to check that a topic exists?

**Comment by blowmage** [comment]
`Gcloud::Pubsub::Project#topic` does make an HTTP request to verify the topic exists. If the topic does not exist it returns `nil` instead of the object. This type of feedback, returning nil instead of the object, is expected by most rubyists. Mistyping the topic name and having a new topic created seems like surprising behavior to me. Its certainly possible to do, but not a behavior I would prefer. Up to you though. :)

**Comment by jgeewax** [comment]
OK, so is there a way to just get a topic object, whether it exists or not? and then publish to it (and maybe get a failure?)

> Mistyping the topic name and having a new topic created seems like surprising behavior to me. Its certainly possible to do, but not a behavior I would prefer. Up to you though. :)

I argued for this elsewhere as well, however the point is a topic is literally nothing more than a name. If there was extra metadata, I could see why we'd want to check, but... there's nothing else.

I'm thinking we should have:

```ruby
project = Gcloud::Pubsub::Project.new 'my-project-id'

# Get a topic and check that it exists:
topic = project.get_topic 'my-topic'

# Get a topic but don't check that it exists:
topic = project.topic 'my-topic'

# Explicitly check that a topic exists later...
topic = project.topic 'my-topic'
topic.exists?

# Get a topic, and set it to auto-create
topic = project.topic 'my-topic' {:auto_create => true}
topic.publish 'my-message'  # Will auto-create the topic if it didn't exist
```

Seem reasonable?

**Comment by blowmage** [comment]
I don't like the idea of having an Topic object that can behave so differently by how it was instantiated. I'd prefer that when I have a Topic object I know it exists. However, I think we can get the same convenience of a weak Topic object with the addition of a `Project#publish` method that takes topic_name as an argument and can auto-create:

```ruby
require "gcloud/pubsub"

pubsub = Gcloud.pubsub "my-project-id"

# Will raise if the topic doesn't exist
pubsub.publish "my-topic", "my-message"
# Will auto-create topic if it doesn't exist
pubsub.publish "my-topic", "my-message", auto_create: true
```

After `publish`, the only thing a Topic object does that Project doesn't is `delete` and `create_subscription`. It doesn't make sense to add `delete` to project, but we can add `create_subscription` using the same auto-create topic logic as `publish`:

```ruby
require "gcloud/pubsub"

pubsub = Gcloud.pubsub "my-project-id"

# Will raise if the topic doesn't exist
pubsub.create_subscription "my-topic", "my-sub"
# Will auto-create topic if it doesn't exist
pubsub.create_subscription "my-topic", "my-sub", auto_create: true
```

Thoughts?

**Comment by jgeewax** [comment]
So the issue is that most people (for this API at least) have been asking for the equivalent of "get_or_create" for topics and subscriptions. That is, they don't want:

```ruby
topic = project.get_topic 'name'
if topic.nil?
  topic = project.create_topic 'name'
end
```

Given this, I'd expect the following behavior:

```ruby
# Get a topic and check that it exists, returning nil if it doesn't.
topic = project.get_topic 'my-topic'
unless topic.nil?
  topic.publish 'my-message'
end

# Get a topic but don't check that it exists, just give back an object.
topic = project.topic 'my-topic'

# We can publish to this topic regardless of whether it exists yet.
topic = project.topic 'my-topic'
topic.publish 'my-message'  # Will create the topic if it didn't exist already.
```

This means we can check for the existence of a topic in two ways:

```ruby
topic = project.get_topic 'name'
exists = !topic.nil?
```

Or

```ruby
topic = project.topic 'name'
exists = topic.exists?
```

**Comment by jgeewax** [comment]
@blowmage Any thoughts?

**Comment by jgeewax** [comment]
@blowmage Can you summarize what we decided today based on the chat? Just want to make sure we're on the same page ....

**Comment by blowmage** [comment]
The plan is move the current Project#topic functionality to Project#get_topic, and have Project#topic return a lazy topic object. [NotFoundError](http://googlecloudplatform.github.io/gcloud-ruby/docs/master/Gcloud/Pubsub/NotFoundError.html) errors raised from the lazy topic object's [`publish` ](http://googlecloudplatform.github.io/gcloud-ruby/docs/master/Gcloud/Pubsub/Topic.html#method-i-publish) will be caught, the topic will be created, and then the publish will be retried. This makes two HTTP calls currently, but in the new best case will only make one HTTP call (although in the new worst case will make three HTTP calls).

**Comment by blowmage** [comment]
@jgeewax We also discussed the possibility of maintaining the object laziness through subscriptions. Suppose we have the following:

```ruby
require "gcloud/pubsub"

pubsub = Gcloud.pubsub
topic = pubsub.topic "i-dont-exist-yet"
sub = topic.subscription "neither-do-i"
msgs = sub.pull
```

The topic object is lazy, and doesn't make an HTTP call when created. Is the subscription object also lazy? If the subscription object is also lazy, and the call to `Subscription#pull` fails because subscription doesn't exist, does it need to be created vs. returning an empty array? When would a lazy subscription ever need to be created?

We could make the `Topic#subscription` method make an HTTP call to see if the subscription exists, and if not make a second HTTP method to create it. But with this option there wouldn't be any lazy subscription objects, only verified objects.

Since we retrieve a subscription only by its name, we also have a [`Project#subscription`](http://googlecloudplatform.github.io/gcloud-ruby/docs/master/Gcloud/Pubsub/Project.html#method-i-subscription) method that doesn't have a topic parameter. We can make that method return a lazy object but it will not be able to create a record since we don't have the topic to create the subscription. But if we never create a missing subscription object that isn't a problem.

```ruby
require "gcloud/pubsub"

pubsub = Gcloud.pubsub
sub = pubsub.subscription "sub-doesnt-exist"
msgs = sub.pull # empty array or raise NotFoundError
sub.acknowledge 1234567890 # raise NotFoundError
sub.topic # nil or raise NotFoundError?
```

The only way to create a lazy subscription object from `Project.subscription` is to change the method signature from `Project.subscription(name)` to `Project.subscription(topic, name)`. But you also said you didn't want a `Project.publish(topic, message)` method because it wasn't OO, so I don't think you'd like a `Project.subscription(topic, name)` method for the same reason.

**Comment by jgeewax** [comment]
@blowmage 

> NotFoundError errors raised from the lazy topic object's publish will be caught, the topic will be created, and then the publish will be retried.

There should be a way to turn this off too (ie, `auto_create: false` or similar)

> When would a lazy subscription ever need to be created?

I think subscriptions need more metadata (ack deadline, and potentially a push configuration) so I don't know if auto-create should be possible there. If we go that route, then we have the same pattern:

- `subscription` gets you a subscription without making an API request. If you call `pull` on it and it turns out that it doesn't exist, we just throw the `NotFound` error
- `get_subscription` will make the API request and return `nil` if `NotFound`.

Note that `subscription` differs from `topic` in that calling `publish` on an unverified topic will auto-create it if the topic doesn't yet exist, whereas calling `pull` on an unverified subscription will raise an exception if the subscription doesn't yet exist.

Sound reasonable?

**Comment by blowmage** [comment]
Yep, we are on the same page. Thanks for the clarification.
